src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py

```diff
@@ -43,6 +49,108 @@ def __init__(self, rseName, **kwargs):
+        :return:           True if read and update were both successful, False otherwise.
+        """
+        # NOTE: The fields to be manipulated are only those which are compatible
+        #       with MongoDB (i.e. here we avoid any field holding a strictly
+        #       pythonic objects, like generators etc.)
+        mongoRecord = collection.find_one(self.mongoFilter)
+
+        # update the list fields read from MongoDB back to strictly pythonic `set`s
+        if mongoRecord:
+            for dirKey, dirList in mongoRecord['dirs'].items():
+                mongoRecord['dirs'][dirKey] = set(dirList)
+            self.update(mongoRecord)
+            return True
+        else:
+            return False
+
+    def writeRSEToMongoDB(self, collection, dumpRSE=False, overwrite=False):
+        """
+        A method to write/update the RSE at the Database from the current object.
+        :param collection: The MonogoDB collection to write on
+        :param dumpRSE:    Bool flag, used to trigger dump of the whole RSE object to
+                           the database with the `files` section (excluding the generator objects)
+                           NOTE: if dumpRSE=False and a previous record for the RSE already exists
+                                 the fields missing from the projection won't be updated
+                                 during this write operation but will preserver their values.
+                                 To completely refresh and RSE record in the database use
+                                 self.purgeRSEAtMongoDB first.
+        :return:           True if update was successful, False otherwise.
+        """
+        # NOTE: The fields to be manipulated are only those which are compatible
+        #       with MongoDB (i.e. here we avoid any field holding a strictly
+        #       pythonic objects, like generators etc.)
+
+        updateFields = {}
+        mongoProjection = self.buildMongoProjection(dumpRSE)
+        for field in mongoProjection:
+            if mongoProjection[field]:
+                # convert back the strictly pythonic `set`s && `generators` to Json compatible lists
+                if field == 'dirs':
+                    updateFields['dirs'] = {}
+                    for dirKey, dirSet in self['dirs'].items():
+                        updateFields['dirs'][dirKey] = list(dirSet)
+                elif field == 'files':
+                    updateFields['files'] = {}
+                    for fileKey, fileSet in self['files'].items():
+                        if isinstance(self['files'][fileKey], dict):
+                            updateFields['files'][fileKey] = list(self['files'][fileKey])
+                        else:
+                            updateFields['files'][fileKey] = self['files'][fileKey]
+                else:
+                    updateFields[field] = self[field]
+
+        updateOps = {'$set': updateFields}
+        if overwrite:
```

Comment:
Almost...

This:
> However, when overwrite=False, a new document does not get created in case it does not exist (it does here because upset=True...).

Is not what is meant in the logic here. `upsert` will always take precedence no mater what the flag of `overwrite` is. `upsert` tells if a missing document is about to be created. `overwrite` relates only to an already existing document. So if a document is missing and the call is telling: "please create the document" it will be created no matter what the other flags stands.
 
And also the update operation is supposed to be working on fields, while the replace operation is supposed  to swap the whole document.